### ISSUE TITLE
Mac OS X uses MAP_ANON instead of MAP_ANONYMOUS

### ISSUE THREAD

**Issue by rednaxelafx** [eclipse/omr#902]
I was trying to play with the JitBuilder examples on Mac OS X 10.9.
Building JitBuilder following instructions from Page 30 of Mark's January OMR JIT talk:
```
$ git clone https://github.com/eclipse/omr
$ cd omr && ./configure [ --disable-warnings-as-errors ]
$ cd jitbuilder && make # builds libjitbuilder.a
$ cd release && make test # builds and runs examples
```
... leads me to this error: 
```
In file included from ../jitbuilder/env/FrontEnd.cpp:30:
../compiler/env/FEBase_t.hpp:94:42: error: use of undeclared identifier 'MAP_ANONYMOUS'
                           MAP_PRIVATE | MAP_ANONYMOUS,
                                         ^
```
I guess we'd have to do conditional compilation here for OS X.

Side question: what degree of support should I expect from OMR/TR on Mac OS X as of now?

**Comment by mstoodle** [comment]
Hey @rednaxelafx ,

Sorry, those directions aren't perfect, I'm afraid. I've got a blog article in the works with the real directions (which include working on OS X, note the SPEC setting):

    $ git clone https://github.com/eclipse/omr
    $ cd omr && make -f run_configure.mk SPEC=osx OMRGLUE=./example/glue
    $ cd jitbuilder && make
    $ cd release && make
    $ make test | testall

I work on OS X myself for all my JitBuilder and OMR development (unless something is Linux-specific), so everything should work just fine on OS X. And if you find something that doesn't work, please open an issue!

**Comment by mstoodle** [comment]
Mind you, I'm running 10.12 ... Please post back here if it still doesn't work with the updated instructions.

**Comment by rednaxelafx** [comment]
Hi Mark @mstoodle ,

Cool! Thanks for your reply.

I did a fresh clone of the `omr` repo and followed your updated instructions. Unfortunately it still fails the same way on my machine.
I wonder if it has anything to do with OS X version that I'm running on. As I said I'm running on a fairly old version, 10.9 (x86_64-apple-darwin13.4.0), and compiling with an old Clang (3.5.x) as well.

Looking forward to your next installment of the JitBuilder blog series!

Best regards,
Kris Mok

**Comment by mstoodle** [comment]
Ok, some research the web suggests there's a simple "fix" which just does some legerdemain with MAP_ANONYMOUS and MAP_ANON. I'll make a patch and then would you be able to try it, Kris @rednaxelafx ?

It's very nice to see you here, by the way! Welcome!

**Comment by mstoodle** [comment]
Hey Kris,
Would you be able to test out the omrjit_osx_map_anonymous branch on my personal fork at https://github.com/mstoodle/omr ? As soon as you verify that resolves the issue for you, I'll submit the pull request.

@rednaxelafx

**Comment by rednaxelafx** [comment]
Hi Mark,

Thanks for the warm welcome ;-)

I tried your patch and it does fix that particular issue. But as the build goes on I'm hitting other issues as well. Apparently my build environment is waaay too old, but since I'll be getting a new laptop soon I'm reluctant to update the system on my current one...

There's another occurrence of the `MAP_ANONYMOUS` issue in `jitbuilder/JBCodeCacheManager.cpp`, and I used the same quick-n-dirty fix in your patch to get around that:
```
diff --git a/jitbuilder/runtime/JBCodeCacheManager.cpp b/jitbuilder/runtime/JBCodeCacheManager.cpp
index 71b084f..d314db7 100644
--- a/jitbuilder/runtime/JBCodeCacheManager.cpp
+++ b/jitbuilder/runtime/JBCodeCacheManager.cpp
@@ -72,6 +72,16 @@ JitBuilder::CodeCacheManager::allocateCodeCacheSegment(size_t segmentSize,
                                               size_t &codeCacheSizeToAllocate,
                                               void *preferredStartAddress)
    {
+
+#if !defined(MAP_ANONYMOUS)
+  #define NO_MAP_ANONYMOUS
+  #if defined(MAP_ANON)
+    #define MAP_ANONYMOUS MAP_ANON
+  #else
+    #define MAP_ANONYMOUS 0
+  #endif
+#endif
+
    // ignore preferredStartAddress for now, since it's NULL anyway
    //   goal would be to allocate code cache segments near the JIT library address
    codeCacheSizeToAllocate = segmentSize;
@@ -85,6 +95,12 @@ JitBuilder::CodeCacheManager::allocateCodeCacheSegment(size_t segmentSize,
                                           0,
                                           0);
 
+// keep the impact of this fix localized
+#if defined(NO_MAP_ANONYMOUS)
+  #undef MAP_ANONYMOUS
+  #undef NO_MAP_ANONYMOUS
+#endif
+
    TR::CodeCacheMemorySegment *memSegment = (TR::CodeCacheMemorySegment *) ((size_t)memorySlab + codeCacheSizeToAllocate - sizeof(TR::CodeCacheMe
    new (memSegment) TR::CodeCacheMemorySegment(memorySlab, reinterpret_cast<uint8_t *>(memSegment));
    return memSegment;
```

And then there's a new error:
```
perl ../jitbuilder/build/scripts/masm2gas.pl --64  -I'../include_core' -I'../jitbuilder/x/amd64' -I'../jitbuilder/x' -I'../jitbuilder' -I'../compiler/x/amd64' -I'../compiler/x' -I'../compiler' -I'..' -o../build/jitbuilder/x/runtime/ ../jitbuilder/x/runtime/AsmUtil64.asm
cannot determine GNU assembler version
Mapping 64-bit registers.
mv ../build/jitbuilder/x/runtime/AsmUtil64.s ../build/jitbuilder/x/runtime/AsmUtil64.o.s
clang -Wa,--noexecstack -march=x86-64 -c  -DTR_HOST_X86=1 -DTR_HOST_64BIT=1 -DBITVECTOR_64BIT=1 -DOSX=1 -DTR_TARGET_X86=1 -DTR_TARGET_64BIT=1 -I'../include_core' -I'../jitbuilder/x/amd64' -I'../jitbuilder/x' -I'../jitbuilder' -I'../compiler/x/amd64' -I'../compiler/x' -I'../compiler' -I'..' -o ../build/jitbuilder/x/runtime/AsmUtil64.o ../build/jitbuilder/x/runtime/AsmUtil64.o.s
../build/jitbuilder/x/runtime/AsmUtil64.o.s:23:8: error: expected identifier after '.ifdef'
.ifdef 1
       ^
```
Apparently `masm2gas.pl` wasn't able to determine my `as` version with:
```
my $gasVersion = (`as --version 2>$null`)[0];
```

Running `as --version` on my Mac OS X 10.9 shows:
```
$ as --version
FATAL:/Library/Developer/CommandLineTools/usr/bin/../libexec/as/x86_64/as: I don't understand '-' flag!
$ as -v
Apple Inc version cctools-862, GNU assembler version 1.38
```
... and it'd be waiting for input from `stdin`. So `-v` has to be used instead of `--version` in my environment, and I'll have to feed some fake input into it to make it terminate after printing the version string.
Just as a hack I did the following to get it recognize the `as`:
```
diff --git a/jitbuilder/build/scripts/masm2gas.pl b/jitbuilder/build/scripts/masm2gas.pl
index 3a84f25..92f3fb2 100644
--- a/jitbuilder/build/scripts/masm2gas.pl
+++ b/jitbuilder/build/scripts/masm2gas.pl
@@ -41,7 +41,7 @@ die "usage: $0 [[-Idir1]* [-M]] file1.asm file2.asm ...\n" unless @ARGV;
 
 $|=1;
 my $null = &devnull;
-my $gasVersion = (`as --version 2>$null`)[0];
+my $gasVersion = (`$null 2>$null | as -v 2>&1`)[0];
 my $llvmVersion = '0';
 
 if ($gasVersion && $gasVersion =~ "LLVM")
@@ -53,7 +53,7 @@ if ($gasVersion && $gasVersion =~ "LLVM")
 elsif ($gasVersion)
    {
    chomp $gasVersion;
-   $gasVersion =~ s/^GNU assembler ([\d\.]+).*$/$1/;
+   $gasVersion =~ s/^.*?GNU assembler version ([\d\.]+).*$/$1/;
    print "using GNU assembler $gasVersion\n";
    }
 else
```
But it doesn't work just yet. The makefile wasn't invoking `as` to assemble that file, but invoking `clang` (LLVM Assembler):
```
clang -Wa,--noexecstack -march=x86-64 -c  -DTR_HOST_X86=1 -DTR_HOST_64BIT=1 -DBITVECTOR_64BIT=1 -DOSX=1 -DTR_TARGET_X86=1 -DTR_TARGET_64BIT=1 -I'../include_core' -I'../jitbuilder/x/amd64' -I'../jitbuilder/x' -I'../jitbuilder' -I'../compiler/x/amd64' -I'../compiler/x' -I'../compiler' -I'..' -o ../build/jitbuilder/x/runtime/AsmUtil64.o ../build/jitbuilder/x/runtime/AsmUtil64.o.s
```
So I actually need to twist the arms in that Perl script to force it generate the files in LLVM Assembler format. I just simply patched the script to assign the correct values...:
```
+$llvmVersion = "3.5";
+$gasVersion = "3.0";
```

And finally I'm able to make `libjitbuilder.a` and the release binary, and then `make test` / `make testall` works fine. Woohoo!

P.S. when building the release binary, the `cp` command wasn't working:
```
cp -u ../compiler/env/defines.h release/include/compiler/env/defines.h || cp ../compiler/env/defines.h release/include/compiler/env/defines.h
cp: illegal option -- u
```

Thanks,
Kris

**Comment by mstoodle** [comment]
Geez, nothing like a smooth onboard experience...Thanks for taking the time to work through all those issues to get to a working state. Annoyed with myself that I didn't at least catch the second MAP_ANONYMOUS, but that's what I get for grepping only in a subdirectory (compiler) :( .

On my system, `as -v` yields the following:

    $ as -v
    Apple LLVM version 8.0.0 (clang-800.0.42.1)
    Target: x86_64-apple-darwin16.4.0
    Thread model: posix
    InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
     ...

and then hangs on stdin. That's the same thing `as --version` prints, but `as --version` is nicer and doesn't wait on stdin. Perhaps there's a solution here, since `as -v < /dev/null` prints roughly the same (superset of) the output from `as --version` on more recent versions of clang.

I'll have to look deeper at the issues in that masm2gas script. I'm glad you were able to hack your way through.

The cp messages I've seen before too (and *thought* I had already fixed...will take care of that tonight for good!). I believe the copy goes ahead, it just complains about the option (otherwise you shouldn't have been able to build the samples).

Again, strong marks for pushing your way through all these problems, and thanks very much for documenting how you did it!

I look forward to hearing any other feedback you have on JitBuilder. The API has been steadily growing and is definitely a work in progress. We're augmenting it as we use it to build out more language JITs as proofs of the concept.

Anyway, I hope your "playing" goes more smoothly from here :) .

**Comment by rednaxelafx** [comment]
Cool! Thank you so much for fixing the build for older versions of OS X so quickly.

For the `masm2gas.pl` issue, is it possibly to hook it up more tightly with the makefiles, e.g. let the makefile pass in arguments to the Perl script to indicate which assembler was chosen as the `S_CMD`. That'll fix the mismatch that I encountered where the Perl script is version-checking `as`, but the makefile is actually using `clang` to assemble.

P.S. I'm having fun reading through the IL dumps from the JitBuilder examples now. I was a bit surprised from the extra local variable store/loads in the initial generated IL, but once I saw the comment in `IlBuilder.cpp` about `TR::IlValue` corresponding to `SymbolReferences`, that made a lot of sense. Although, I might be more willing to generate TR IL directly, that way it gives me fine-grain control over what the initial IL would look like.

**Comment by mstoodle** [comment]
> P.S. I'm having fun reading through the IL dumps from the JitBuilder examples now. I was a bit surprised from the extra local variable store/loads in the initial generated IL, but once I saw the comment in IlBuilder.cpp about TR::IlValue corresponding to SymbolReferences, that made a lot of sense. Although, I might be more willing to generate TR IL directly, that way it gives me fine-grain control over what the initial IL would look like.

Glad you're having fun :) .

Yeah, JitBuilder tries to insulate you from details of the IL and relies on the first few optimizations just to clean things up. The IL generation from JitBuilder is *brain-dead* simple. There's an issue opened already (Issue #834) to improve the quality of the initial IL with some initial thoughts.

The superclass of IlBuilder, called IlInjector, was intended to be an intermediate interface to the compiler IL (I actually built that first and it's used in some of the compiler fvtest suite; some of the experiences people had working with it led me to develop the IlBuilder layer). You have to know what you're doing, but some of the detailed mechanics can be avoided. That API is pretty barebones right now; if you're interested in that layer, I'd be happy to discuss how it could be improved.

**Comment by mstoodle** [comment]
Hey, @rednaxelafx , just wanted to follow up here. Since #1110 was merged, the initial IL generated for JitBuilder should be a lot easier to read. And once #1146 is merged, it should look very much like the incoming IlBuilder code after just the first optimization (dead trees elimination). Of course, you're perfectly welcome to generate your own IL :) .

**Comment by mstoodle** [comment]
Since the original problems you opened to report your issues is addressed now, would you agree we can close this issue @rednaxelafx ? We can copy your issues with masm2gas.pl into another issue, perhaps?
